### DISCUSSION TITLE
How add folders to 1 exe files

### DISCUSSION THREAD

**Discussion by KOKASGKASKGO125** [pyinstaller/pyinstaller#7822] [Help] [answered]
my project

```text
main.py
templates -
     - index.html
     - test.html
     - ..... .html
static-
     - css
         -  index.css
    - js
         - index.js
```
my packing command 
```bash
pyinstaller --onefile --add-data "./static/*;static" --add-data "./templates/*;templates" main.py
```

result

```python
Traceback (most recent call last):
  File "main.py", line 1, in <module>
  File "<frozen importlib._bootstrap>", line 1176, in _find_and_load
  File "<frozen importlib._bootstrap>", line 1147, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 690, in _load_unlocked
  File "PyInstaller\loader\pyimod02_importers.py", line 385, in exec_module
  File "app.py", line 19, in <module>
  File "starlette\staticfiles.py", line 57, in __init__
RuntimeError: Directory 'static' does not exist
[4600] Failed to execute script 'main' due to unhandled exception!`
```

**Comment by bwoodsend** [comment]
`--add-data static;static` is the correct syntax for PyInstaller. Make sure you are referencing them correctly in your Python code though. `open("static/js/index.js")` is not valid (see [docs](https://pyinstaller.org/en/stable/runtime-information.html#using-file)).

**Reply by KOKASGKASKGO125** [reply]
js files are loaded in index.html as in a normal html page

**Reply by KOKASGKASKGO125** [reply]
Everything works without assembly, after assembly it cannot find the folder, I no longer know how to solve it

**Reply by bwoodsend** [reply]
How are you referencing these files in your code? I can't do anything without seeing some code.

**Reply by KOKASGKASKGO125** [reply]
I have templates and static from jinja2 connected `templates = Jinja2Templates(directory="templates")
app.mount('/static', StaticFiles(directory='static', html=True), name='static')`

**Reply by bwoodsend** [reply]
Ughh yuck. jinja2 tries to be clever and manage resource location for you which in practice means that it chooses somewhere completely nonsensical, won't tell you where and won't let configure it. Let me see where it's arbitrarily assigned this time...

**Reply by KOKASGKASKGO125** [reply]
I use jinja2 

```python
app = FastAPI()

templates = Jinja2Templates(directory="templates")
app.mount('/static', StaticFiles(directory='static', html=True), name='static')

@app.get("/", response_class=HTMLResponse)
async def read_root(request: Request):
    return templates.TemplateResponse("index.html", {"request": request})
```

**Reply by bwoodsend** [reply]
Can you include the imports - I see Jinja2Templates isn't actually part of jinja2 so I don't know where it comes from.

**Reply by bwoodsend** [reply, marked as answer]
This works for me.
```python
from pathlib import Path

...

templates = Jinja2Templates(directory=str(Path(__file__).with_name("templates")))
app.mount('/static', StaticFiles(directory=str(Path(__file__).with_name('static')), html=True), name='static')
```

**Reply by Lopou94** [reply]
Hello,
I have created a project with an ADBT.py file which is the main graphics of my program and an ADBD.py which is my program definitions.
I have built the whole project in this path "**C:\Users\papostolopoulos\PycharmProjects\pythonProject\ADB**".
My software needs the platform-tools (ADB) and scrcpy programs to work, which are stored in the path "**C:\Users\NAMEOFPYPC\PycharmProjects\pythonProject\ADB**" in the platform-tools and scrcpy folders.
I tried with this command:
**"pyinstaller --noconsole --onefile --add-data "C:\\Users\\papostolopoulos\\PycharmProjects\\pythonProject\\ADB\\platform-tools;platform-tools" --add-data "C :\\Users\\papostolopoulos\\PycharmProjects\\pythonProject\\ADB\\scrcpy;scrcpy" ADBT.py**
" to merge these two folders, but I'm facing this issue:

**21633 INFO: checking PYZ
21633 INFO: Building PYZ because PYZ-00.toc is non existent
21633 INFO: Building PYZ (ZlibArchive) C:\Users\papostolopoulos\PycharmProjects\pythonProject\ADB\build\ADBT\PYZ-00.pyz
22821 INFO: Building PYZ (ZlibArchive) C:\Users\papostolopoulos\PycharmProjects\pythonProject\ADB\build\ADBT\PYZ-00.pyz completed successfully.
22875 INFO: checking PKG
22876 INFO: Building PKG because PKG-00.toc is non existent
22876 INFO: Building PKG (CArchive) ADBT.pkg
37254 INFO: Building PKG (CArchive) ADBT.pkg completed successfully.
37325 INFO: Bootloader C:\Users\papostolopoulos\PycharmProjects\pythonProject\.venv\Lib\site-packages\PyInstaller\bootloader\Windows-64bit-intel\runw.exe
37362 INFO: checking EXE
37369 INFO: Building EXE because EXE-00.toc is non existent
37369 INFO: Building EXE from EXE-00.toc
37370 INFO: Copying bootloader EXE to C:\Users\papostolopoulos\PycharmProjects\pythonProject\ADB\dist\ADBT.exe
37383 INFO: Copying icon to EXE
37493 INFO: Copying 0 resources to EXE
37496 INFO: Embedding manifest in EXE
37622 WARNING: Execution of 'write_manifest_to_executable' failed on attempt #1 / 20: error(5, 'LoadLibraryExW', 'Access is denied.'). Retrying in 0.05 second(s)...
47743 INFO: Appending PKG archive to EXE
47857 INFO: Fixing EXE headers
48660 INFO: Building EXE from EXE-00.toc completed successfully.**

Can I include these folders in my .exe file with the way you said?

Thanks in advance,

**Reply by rokm** [reply]
@Lopou94 What exactly is "this issue"? If you are referring to the WARNING line, that's likely caused by your anti-virus program interfering with the build.

**Reply by YaoJusheng** [reply]
> 
> ```python
> from pathlib import Path
> 
> ...
> 
> templates = Jinja2Templates(directory=str(Path(__file__).with_name("templates")))
> app.mount('/static', StaticFiles(directory=str(Path(__file__).with_name('static')), html=True), name='static')
> ```

I'm having a similar problem, but I can't solve it with a simple `Path(__file__).with_name(...)`.

The structure of the project is as follows:

```
.
├── alembic.ini
├── app
│   ├── api/
│   │   └── ...
│   ├── config.py
│   ├── core
│   │   ├── context
│   │   │   ├── __init__.py
│   │   │   └── request.py
│   │   ├── extension
│   │   │   ├── __init__.py
│   │   │   └── mailing.py
│   ├── static
│   │   ├── js
│   │   ├── css
│   │   ├── index.html
│   │   └── ...
│   ├── templates
│   │   └── ...
│   ├── translations
│   │   └── ...
│   ├── __init__.py
│   └── ...

```

in `config.py`
```python

class Config:
	ROOT_DIR  = Path(os.getcwd())
    BASE_DIR = ROOT_DIR / "app"
    STUB_DIR = ROOT_DIR / ".stub"
    STATIC_FOLDER = BASE_DIR / "static"
    TEMPLATE_FOLDER = BASE_DIR / "templates"
    LOCALES_FOLDER = BASE_DIR / "translations"
    ...
```

in `app/core/context/request.py`
```python
templates = Jinja2Templates(directory=Config.TEMPLATE_FOLDER)

@asynccontextmanager
async def lifespan(app: FastAPI):
	templates.env.globals["_"] = translator._
	...
```

in` app/core/extension/mailing.py`
```python
mail_conf = MailConnectionConfig(
    MAIL_USERNAME=Config.MAIL_USERNAME,
    MAIL_PASSWORD=Config.MAIL_PASSWORD,
    MAIL_FROM=Config.MAIL_FROM,
    MAIL_SERVER=Config.MAIL_SERVER,
    MAIL_PORT=Config.MAIL_PORT,
    MAIL_FROM_NAME=Config.MAIL_FROM_NAME,
    MAIL_STARTTLS=Config.MAIL_STARTTLS,
    TEMPLATE_FOLDER=Config.TEMPLATE_FOLDER,
)
mail_client: FastMail = FastMail(mail_conf)
```

in `app/__init__.py`
```python
app = FastAPI(lifespan=lifespan)

app.mount(
    "/static",
    StaticFiles(directory=Config.STATIC_FOLDER),
    name="static",
)

@app.get("/")
async def index(request: Request):
    return HTMLResponse(content=(Config.STATIC_FOLDER / "index.html").read_text("utf-8"))
```

When I use command like:
```bash
pyinstaller \
  --onefile manage.py \
  --log-level=WARN \
  --noconfirm \
  --clean \
  --add-data "app:app" \
  --collect-all aiosqlite \
  --collect-all passlib   \
  --collect-all ruamel.yaml \
  --distpath pydist \
  --name manage
```
 and startup with command:
```
./manage 
```
then output error:
```
FileNotFoundError: [Errno 2] No such file or directory: '/home/coolker/dist/app/static/index.html'
```

**Reply by rokm** [reply]
> I'm having a similar problem, but I can't solve it with a simple `Path(__file__).with_name(...)`.

You should also set your `Config.ROOT_DIR` to point to `Path(__file__)` - currently, you have it set to current working directory...

**Reply by YaoJusheng** [reply]
> > I'm having a similar problem, but I can't solve it with a simple `Path(__file__).with_name(...)`.
> 
> You should also set your `Config.ROOT_DIR` to point to `Path(__file__)` - currently, you have it set to current working directory...

`config.py` is not in the root directory. The purpose of setting `ROOT_DIR` is to find the working directory where the program runs.

I don't understand why in this way, the data files cannot be packaged into a binary file.

It seems that `pyinstaller` cannot correctly handle path like this during the compilation and packaging process?

**Reply by rokm** [reply]
But in the PyInstaller-frozen application, your resources are NOT in the current working directory. And neither would they be if you run your `manage.py` script unfrozen from an arbitrary working directory (e.g., `cd / && python /path/to/project/manage.py`).

**Reply by YaoJusheng** [reply]
> But in the PyInstaller-frozen application, your resources are NOT in the current working directory. And neither would they be if you run your `manage.py` script unfrozen from an arbitrary working directory (e.g., `cd / && python /path/to/project/manage.py`).

OK, thank you very much for your instant reply, I seem to understand what you said. 

It seems that I need to rethink the path in the project, but I still have some doubts, that is, why can't the resource path be found when executing in onefile mode in the parent directory of the binary file packaged in pydist (set through compilation options).

**Comment by KOKASGKASKGO125** [comment]
> Can you include the imports - I see Jinja2Templates isn't actually part of jinja2 so I don't know where it comes from. [Можете ли вы включить импорт - я вижу, что Jinja2Templates на самом деле не является частью jinja2, поэтому я не знаю, откуда он взялся.]

```python

from fastapi import FastAPI, Request, UploadFile, File, Query
from fastapi.templating import Jinja2Templates
from fastapi.staticfiles import StaticFiles
from fastapi.responses import HTMLResponse

```
